src/Adapter/DoctrineItems.php

```diff
@@ -0,0 +1,354 @@
+    {
+        if (null !== $cursor && ((string) (int) $cursor) !== $cursor) {
+            return new ItemListPage([], null);
+        }
+
+        $query = $this->connection->createQueryBuilder();
+
+        $query
+            ->addSelect('item.id')
+            ->addSelect('item.sequence')
+            ->from($this->tableItems, 'item')
+            ->orderBy('item.sequence')
+            ->setMaxResults($limit + 1);
+
+        if ($cursor) {
+            $query
+                ->where($query->expr()->gte('item.sequence', ':cursor'))
+                ->setParameter('cursor', (int) $cursor);
+        }
+
+        $results = $query->execute();
+        $ids = $results->fetchAll(FetchMode::ASSOCIATIVE);
+
+        if (empty($ids)) {
+            return new ItemListPage([], null);
+        }
+
+        if (count($ids) > $limit) {
+            $newCursor = (string) array_pop($ids)['sequence'];
+        }
+
+        $ids = array_map([ItemId::class, 'fromString'], array_column($ids, 'id'));
+
+        return new ItemListPage($ids, $newCursor ?? null);
+    }
+
+    public function count() : int
+    {
+        $results = $this->connection->executeQuery("SELECT COUNT(item.id) count FROM {$this->tableItems} item");
+
+        return (int) $results->fetch()['count'];
+    }
+
+    public function getIterator() : Traversable
+    {
+        while (true) {
+            /** @var Statement $statement */
+            $results = $this->connection->executeQuery(
+                'SELECT version.id, version.version, version.content, version.hash, item.sequence'
+                ." FROM {$this->tableItems} item LEFT JOIN 
+{$this->tableVersions} version ON item.id = version.id AND item.version = version.version AND item.sequence > :sequence"
+                .' ORDER BY item.sequence LIMIT 100',
+                ['sequence' => $cursor ?? 0]
+            );
+
+            while ($result = $results->fetch(FetchMode::ASSOCIATIVE)) {
+                yield $this->rowToItemVersion($result);
+                $cursor = $result['sequence'];
+            }
+
+            if (empty($result)) {
+                break;
+            }
+        }
+    }
+
+    public function getSchema() : Schema
+    {
+        $schema = new Schema();
+
+        $table = $schema->createTable($this->tableItems);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('sequence', Type::INTEGER, ['autoincrement' => true]);
+        $table->setPrimaryKey(['sequence']);
+        $table->addUniqueIndex(['id']);
+
+        $table = $schema->createTable($this->tableVersions);
+        $table->addColumn('id', Type::STRING);
+        $table->addColumn('version', Type::INTEGER);
+        $table->addColumn('content', Type::BLOB);
+        $table->addColumn('hash', Type::STRING);
+        $table->setPrimaryKey(['id', 'version']);
+        $table->addIndex(['id']);
```

Comment:
in general RDBMS usually can use the composite index of the primary key `(id, version)` to also satisfy queries oriented to just `id` (any prefix of fields), but not sure if it's part of the standard. In that case, this additional index would be unnecessary

Reply:
Trying to work out what would use this. _Might_ only be https://github.com/libero/content-api-bundle/blob/82d617bcfb7d265e6c785377068681357ba15625/src/Adapter/DoctrineItems.php#L274

🤔

Reply:
So if it comes for free from the other index, and it might be rarely used anyway, all the reason not to have it